Can't open attached PDF of message in extern application

<!--
This is a bug report tracker. New features are discussed in the forum: https://support.delta.chat

Please fill out as much of this form as you can (leaving out stuff that is not applicable is ok).
-->

- Android version: 14
- Device: Ulefone 25T Pro
- Delta Chat version: 1.50.3
- Expected behavior: when the PDF opens in my PDF Viewer, it should get shown
- Actual behavior: the PDF Viewer opens but do show nothing, a other one crash
- Steps to reproduce the problem: open any send or received PDF
- Screenshots:
- Logs:

<!--
Debug logs can be copied from within the Delta Chat app with
Settings menu -> Advanced -> View log

Alternatively from the Android system log:
`adb logcat -v time -s DeltaChat`

Logs may contain private data 
which shall be removed or anonymised prior to posting.
-->

with "Libreara FD" it works, but I installed "Collabora Office" and in fact it fails to open the PDF from Delta Chat while from Telegram and WhatsApp it works, here is the log:

```
FATAL EXCEPTION: main
                                                                                                    Process: com.collabora.libreoffice, PID: 26415
                                                                                                    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.collabora.libreoffice/org.libreoffice.androidlib.LOActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
                                                                                                    	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4129)
                                                                                                    	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4316)
                                                                                                    	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:222)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeNonLifecycleItem(TransactionExecutor.java:133)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeTransactionItems(TransactionExecutor.java:103)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:80)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2719)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:109)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:232)
                                                                                                    	at android.os.Looper.loop(Looper.java:317)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8787)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:591)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:871)
                                                                                                    Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
                                                                                                    	at org.libreoffice.androidlib.LOActivity.canDocumentBeExported(LOActivity.java:1233)
                                                                                                    	at org.libreoffice.androidlib.LOActivity.initUI(LOActivity.java:348)
                                                                                                    	at org.libreoffice.androidlib.LOActivity.init(LOActivity.java:258)
                                                                                                    	at org.libreoffice.androidlib.LOActivity.onCreate(LOActivity.java:251)
                                                                                                    	at android.app.Activity.performCreate(Activity.java:9019)
                                                                                                    	at android.app.Activity.performCreate(Activity.java:8997)
                                                                                                    	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1528)
                                                                                                    	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4111)
                                                                                                    	... 13 more
```

tbh, maybe we could do something different to match more what the other apps do but it looks like a bug in the app, some other PDF viewers work fine

--- a/src/androidTest/java/com/b44t/messenger/uitests/offline/SharingTest.java
+++ b/src/androidTest/java/com/b44t/messenger/uitests/offline/SharingTest.java
@@ -93,7 +93,7 @@ public class SharingTest {
       }
     }
     Uri uri = Uri.parse("content://" + BuildConfig.APPLICATION_ID + ".attachments/" + Uri.encode(pngImage));
-    DcHelper.sharedFiles.put(pngImage, 1);
+    DcHelper.sharedFiles.put(pngImage, "image/png");
 
     Intent i = new Intent(Intent.ACTION_SEND);
     i.setType("image/png");

--- a/src/main/java/org/thoughtcrime/securesms/connect/AttachmentsContentProvider.java
+++ b/src/main/java/org/thoughtcrime/securesms/connect/AttachmentsContentProvider.java
@@ -5,8 +5,14 @@ import android.content.ContentValues;
 import android.database.Cursor;
 import android.net.Uri;
 import android.os.ParcelFileDescriptor;
+import android.webkit.MimeTypeMap;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 
 import com.b44t.messenger.DcContext;
+
+import org.thoughtcrime.securesms.util.MediaUtil;
 
 import java.io.File;
 import java.io.FileNotFoundException;
@@ -34,12 +40,12 @@ public class AttachmentsContentProvider extends ContentProvider {
         // where ef39a39 is the file in the blob directory
         // and text.txt is the original name of the file, as returned by `msg.getFilename()`.
         // `uri.getPathSegments()` returns ["ef39a39", "text.txt"] in this example.
-        String path = uri.getPathSegments().get(0);
-        if (!DcHelper.sharedFiles.containsKey(path)) {
+        String file = uri.getPathSegments().get(0);
+        if (!DcHelper.sharedFiles.containsKey(file)) {
             throw new FileNotFoundException("File was not shared before.");
         }
 
-        File privateFile = new File(dcContext.getBlobdir(), path);
+        File privateFile = new File(dcContext.getBlobdir(), file);
         return ParcelFileDescriptor.open(privateFile, ParcelFileDescriptor.MODE_READ_ONLY);
     }
 
@@ -49,8 +55,17 @@ public class AttachmentsContentProvider extends ContentProvider {
     }
 
     @Override
-    public String getType(Uri arg0) {
-        return null;
+    public String getType(Uri uri) {
+        String file = uri.getPathSegments().get(0);
+        String mimeType = DcHelper.sharedFiles.get(file);
+
+        return DcHelper.checkMime(uri.toString(), mimeType);
+    }
+
+    @Override
+    public String getTypeAnonymous(Uri uri) {
+        String ext = MediaUtil.getFileExtensionFromUrl(uri.toString());
+        return MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext);
     }
 
     @Override

--- a/src/main/java/org/thoughtcrime/securesms/connect/DcHelper.java
+++ b/src/main/java/org/thoughtcrime/securesms/connect/DcHelper.java
@@ -266,7 +266,9 @@ public class DcHelper {
     return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
   }
 
-  public static final HashMap<String, Integer> sharedFiles = new HashMap<>();
+  // When the user shares a file to another app or opens a file in another app, it is added here.
+  // `HashMap<file, mimetype>` where `file` is the name of the file in the blobdir (not the user-visible filename).
+  public static final HashMap<String, String> sharedFiles = new HashMap<>();
 
   public static void openForViewOrShare(Context activity, int msg_id, String cmd) {
     DcContext dcContext = getContext(activity);
@@ -291,12 +293,13 @@ public class DcHelper {
       }
 
       Uri uri;
+      mimeType = checkMime(filename, mimeType);
       if (path.startsWith(dcContext.getBlobdir())) {
         // Build a Uri that will later be passed to AttachmentsContentProvider.openFile().
         // The last part needs to be `filename`, i.e. the original, user-visible name of the file,
         // so that the external apps show the name of the file correctly.
         uri = Uri.parse("content://" + BuildConfig.APPLICATION_ID + ".attachments/" + Uri.encode(file.getName()) + "/" + Uri.encode(filename));
-        sharedFiles.put(file.getName(), 1); // as different Android version handle uris in putExtra differently, we also check them on our own
+        sharedFiles.put(file.getName(), mimeType); // as different Android version handle uris in putExtra differently, we also check them on our own
       } else {
         if (Build.VERSION.SDK_INT >= 24) {
           uri = FileProvider.getUriForFile(activity, BuildConfig.APPLICATION_ID + ".fileprovider", file);
@@ -306,7 +309,6 @@ public class DcHelper {
       }
 
       if (cmd.equals(Intent.ACTION_VIEW)) {
-        mimeType = checkMime(filename, mimeType);
         Intent intent = new Intent(Intent.ACTION_VIEW);
         intent.setDataAndType(uri, mimeType);
         intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
@@ -357,7 +359,7 @@ public class DcHelper {
     activity.startActivity(intent);
   }
 
-  private static String checkMime(String path, String mimeType) {
+  public static String checkMime(String path, String mimeType) {
     if(mimeType == null || mimeType.equals("application/octet-stream")) {
       path = path.replaceAll(" ", "");
       String extension = MediaUtil.getFileExtensionFromUrl(path);

--- a/src/main/java/org/thoughtcrime/securesms/connect/DcHelper.java
+++ b/src/main/java/org/thoughtcrime/securesms/connect/DcHelper.java
@@ -299,7 +299,11 @@ public class DcHelper {
         // The last part needs to be `filename`, i.e. the original, user-visible name of the file,
         // so that the external apps show the name of the file correctly.
         uri = Uri.parse("content://" + BuildConfig.APPLICATION_ID + ".attachments/" + Uri.encode(file.getName()) + "/" + Uri.encode(filename));
-        sharedFiles.put(file.getName(), mimeType); // as different Android version handle uris in putExtra differently, we also check them on our own
+
+        // As different Android version handle uris in putExtra differently,
+        // we also check on our own that the file was actually shared.
+        // The check happens in AttachmentsContentProvider.openFile().
+        sharedFiles.put(file.getName(), mimeType);
       } else {
         if (Build.VERSION.SDK_INT >= 24) {
           uri = FileProvider.getUriForFile(activity, BuildConfig.APPLICATION_ID + ".fileprovider", file);

